### PR TITLE
Add git branch naming strategy

### PR DESCRIPTION
**Pull request by bohme**
### Git Branch Naming Strategy

In order to help differentiate git branches a little easier prefix your branch with a relevant keyword.

For example:

| Keyword | Meaning                                             |
|---------|-----------------------------------------------------|
| feature | new additions                                       |
| fix     | fix/adjustment                                      |
| style   | Style changes                                       |
| spike   | experiements & exploration                          |
| meta    | changes to credentials, .gitignore, rake tasks, etc |

### CODE REVIEW

**Comment by GeoffAbtion** [comment]
Convince me that `test` needs its own category.

**Comment by bohme** [comment]
The idea is to use it for branches that only add test.
so we can have things like this:
`test/that-model-over-there`

**Comment by bohme** [comment]
It's also worth noting that these are not set in stone at all.
If this even belongs here.

**Comment by GeoffAbtion** [comment]
> The idea is to use it for branches that only add test.
> so we can have things like this:
> `test/that-model-over-there`

I'd use `fix` for that.
But I guess it doesn't really matter, people can use what they want.

File: README.md [hunk 1]
```diff
@@ -20,3 +20,15 @@ We use Heroku pipelines for managing production and staging environments.
 Heroku-Postgresql databses include *FREE* scheduled backups, we just need to enable them using:
 
 `heroku pg:backups:schedule DATABASE_URL --at '03:00 Europe/Copenhagen' --app APP_NAME`
+
+### Git Branch Naming Strategy
+
+In order to help differentiate git branches a little easier prefix your branch with a relevant keyword.
+
+For example:
```
Comment (hcarreras) [inline README.md]: This is not an example. It's a company-wise guideline :)

File: README.md [hunk 2]
```diff
@@ -25,10 +25,8 @@ Heroku-Postgresql databses include *FREE* scheduled backups, we just need to ena
 
 In order to help differentiate git branches a little easier prefix your branch with a relevant keyword.
 
-For example:
-
 | Keyword | Meaning        |
 | ------- | -------------- |
 | feature | new addition   |
 | fix     | fix/adjustment |
-| style   | Style changes  |
```
Comment (GeoffAbtion) [inline README.md]: Can we keep `style`?

I would expect it to be used for changes that did not change the functionality of the application, but changed how it looks.

**Comment by GeoffAbtion** [comment]
To add: `meta` - For changes unrelated to the app.

This would be things like changing CI settings or `.gitignore`s or adding a rake task to generate documentation.

File: README.md [hunk 3]
```diff
@@ -20,3 +20,15 @@ We use Heroku pipelines for managing production and staging environments.
...
+### Git Branch Naming Strategy
+
+In order to help differentiate git branches a little easier prefix your branch with a relevant keyword.
+
+| Keyword | Meaning                                             |
+|---------|-----------------------------------------------------|
+| feature | new additions                                       |
+| fix     | fix/adjustment                                      |
+| style   | Style changes                                       |
+| spike   | experiements & exploration                          |
```
Comment (hcarreras) [inline README.md]: I think you misspelled `experiments`.

**Comment by hcarreras** [comment]
Don't forget to `squash` before merging.
Also, let's revisit this in a few weeks. I'm unsure if `fix` is a good name for adding tests. It's not self-explanatory and it might be misleading.

**Comment by bohme** [comment]
`fix` isn't just for tests, it's more for code that, well, _fixes_, things.
Be it adding specs or fixing typos or even some adjustments to broken logic.
Generally if it improves the code without adding a new functionality.

**Comment by hcarreras** [comment]
I don't understand `fix` as "improves something". Rather, I understand "it makes work something that wasn't working before". Not that my definition is the correct one, but it would be misleading to me if I take over a branch named `fix` and has a bunch of tests. Immediately, I would wonder why the tests are passing and I would waste some time researching.

**Comment by bohme** [comment]
I'd argue that if you fix something that didn't work before you are _improving something_ 🤔 
In the same spirit that a PR just adding specs fixes hardship for people working on the project and the project itself.

More importantly: if we add too many prefixes people wont use them.
so getting too granular will probably backfire.
so:

feature; general additions (generally new functionality, review might require an extra careful look)
fix: adjustments (the most _umbrella_ of prefixes)
style: pure style changes (real handy for review)
spike: experiments (these branches should never be merged)
meta:  changes to repo meta information (CI, gitignore, etc)

90% of all branches will probably have either the `feature` or `fix` prefix 🤔

**Comment by hcarreras** [comment]
My point is not that my definition is correct, just that it would mislead me.

I agree that most of the branches will be `feature` or  `fix`. In my opinion, `test` is more important than `spike` or `meta`, so if we are afraid that there are too many prefixes, those two could be merged.

If you disagree with me, it's okay. I prefer to revisit this when we have some information about what people think and how they use them.

**Comment by bohme** [comment]
what would `test` be for?
for PRs just adding tests? use `fix` :3
for PRs testing out something new? use `spike` :3

I think 5 is a good number, 2 that will get used a lot, and 3 for special cases that affect reviews.

I agree that `fix`s role is a little blurry or just plain doing too much.
But I have yet to find a convenient way of resolving this.

**Comment by heatherm** [comment]
I would like to see chore in here.

**Comment by bohme** [comment]
what would a `chore` branch look like or do?
updating gems? bumping ruby versions? that kinda thing?

**Comment by DeWinther** [comment]
Would it be a good idea to include e.g. if the work is frontend or backend related? 
I have personally tried doing e.g. 
`feature/fe/some_frontend_implementation` 
`feature/be/some_backend_implementation` 
I like doing this more than having the `style` branch naming

**Comment by reinisla** [comment]
When writing tests I usually have `spec/` prefix.
For style changes, I have`design/` as it is not only CSS that is usually changed.

I'd prefer `test` over `fix` for tests. For me `fix` is for bugfixes or maybe some text changes.

**Comment by GeoffAbtion** [comment]
Tests shouldn't be a separate thing 😠

**Comment by jeppester** [comment]
I'm not convinced that this is necessary, at least not for the sake of "help differentiate git branches"

Branches are created and deleted all the time, and you rarely ever use the name for anything, other than your own personal convenience - which is a good reason to not give your branch a completely useless name.

One real benefit I can see from this, is that it might nudge developers towards not doing too much in their PR's which I think is a good goal, but we could also just tell people to "Aim towards doing only a single thing in your PR".

**Comment by jeppester** [comment]
To add to the above comment:
It's not like I don't want naming conventions anywhere, or conventions in general, I just think each convention should have an clear benefit over the added annoyance of having to learn and adhere to yet another convention.

There are other places where I think conventions could be more useful:
- PR's (naming, description, tags)
- Commit messages

**Comment by bohme** [comment]
Jesper is making some very valid points here 👍 
These prefixes are mostly gonna be to the benefit of the reviewer.

as to `design` over `style` we have been jumping back and forth on that one:
bd42146542be3c2bb8f560b84013efe81423c352

**Comment by jeppester** [comment]
If we want to make code easier to review, then I would argue that a convention for PRs is a better place to start.

**Comment by heatherm** [comment]
> what would a `chore` branch look like or do?
> updating gems? bumping ruby versions? that kinda thing?

This would be any work that must be accepted by a developer. Background tasks that have to be done that has an impact of several features. Example: adding delayed job. The purpose of naming them chore is to be more explicit about the level of testing to be done by the dev team during review.

File: README.md [hunk 4]
```diff
@@ -20,3 +20,15 @@ We use Heroku pipelines for managing production and staging environments.
...
+
+In order to help differentiate git branches a little easier prefix your branch with a relevant keyword.
+
+| Keyword | Meaning                                             |
+|---------|-----------------------------------------------------|
+| feature | new additions                                       |
+| fix     | fix/adjustment                                      |
+| style   | Style changes                                       |
+| spike   | experiments & exploration                           |
+| meta    | changes to credentials, .gitignore, rake tasks, etc |
```
Comment (heatherm) [inline README.md]:
```suggestion
| meta    | changes to credentials, .gitignore, rake tasks, etc |
| chore   | background tasks that must be approved by a dev|
```

Comment (bohme) [reply README.md]: We should be careful not to add too many options else people wont use them.
While still not a fan of the name I do see it's usefulness 🤔

**Comment by hcarreras** [comment]
This discussion became stalled. 
I'm closing this as the team does not seem to agree on the value. Feel free to reopen with a clear description of the value.

**Comment by bohme** [comment]
I won't be reopening this 💃 
No way to  get everyone to agree on unified branch prefixes, I, like most, will keep using the ones we use and new people will just have to pick the ones they likes ^_^

**Pull request closed without merging**